src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -93,10 +100,16 @@ internal AssemblyLoadContext(bool fRepresentsTPALoadContext, bool isCollectible)
 
         ~AssemblyLoadContext()
         {
-            // Only valid for a Collectible ALC. Non-collectible ALCs have the finalizer suppressed.
-            // We get here only in case the explicit Unload was not initiated.
-            Debug.Assert(state != InternalState.Unloading);
-            InitiateUnload();
+            // Use the unloadLock as a guard to detect the corner case when the constructor of the AssemblyLoadContext was not executed
+            // e.g. due to the JIT failing to JIT it.
+            if (unloadLock != null)
```

Comment:
Nit: While you are on it - could you please change `unloadLock` to `_unloadLock` to make it match the coding conventions